File: others/preface/src/regular_number.rb
#one queue solution
def get_number(n)
  q = [1]
  while n>0 do
    x = q.shift
    q = bin_unique_enqueue(q, 2*x)
    q = bin_unique_enqueue(q, 3*x)
    q = bin_unique_enqueue(q, 5*x)
    #q = unique_enqueue(q, 2*x)
    #q = unique_enqueue(q, 3*x)
    #q = unique_enqueue(q, 5*x)
    n -= 1
  end
  return x
end
#trivial unique enqueue
def unique_enqueue(q, x)
  l = 0
  u = q.size - 1
  while l <= u
    if q[l] < x
      l += 1
    elsif q[l] > x
      break
    else
      return q
    end
  end
  q.insert(l, x)
end
#binary search to enqueue
def bin_unique_enqueue(q, x)
  l = 0
  u = q.size-1
  while l <= u
    mid = (u+l) >> 1
    if q[mid] < x
      l = mid + 1
    elsif q[mid] > x
      u = mid - 1
    else
      return q
    end
  end
  q.insert(l, x)
end

#three queue solution
def get_number2(n)
  return 1 if n == 1
  q1 = [2]
  q2 = [3]
  q3 = [5]
  while n>1 do
    x = [q1.first, q2.first, q3.first].min
    if x == q1.first
      q1.shift
      q1 << 2*x
      q2 << 3*x
      q3 << 5*x
    elsif x ==  q2.first
      q2.shift
      q2 << 3*x
      q3 << 5*x
    else
      q3.shift
      q3 << 5*x
    end
    n -= 1
  end
  return x
end

#
# get_number with unique_enqueue:     0.020797470000000012
# get_number with bin_unique_enqueue: 0.0024227199999999997
# get_numerb2:                        0.0007246699999999995
#
def test
  tm = 0
  ret = 0
  100.times do
    start = Time.now
    #ret = get_number(1500)
    ret = get_number2(1500)
    stop = Time.now
    tm += (stop-start)
  end
  puts "average: #{tm/100} s. ret = #{ret}"
end

test()

File: others/preface/src/regular_number.clj
(ns regular-number)

;;trivial solution
(defn recur-mod [x n]
  (if (zero? (rem x n))
    (recur (quot x n) n)
    x))

(defn rns1 []
  (filter
   (fn [x]
     (= 1
        (-> x
            (recur-mod 2)
            (recur-mod 3)
            (recur-mod 5))))
   (iterate inc 1)))

;;one queue
(defn unique-insert [v x]
  (if (some #{x} v)
    v
    (let [[l u] (split-with #(< % x) v)]
      (apply conj (vec l) x u))))

(defn rns2
  {:test (fn []
           (assert
            (=
             859963392
             (time (rns2 1500)))))}
  [n]
  (loop [m (dec n)
         q [1]]
    (if (zero? m)
      (first q)
      (recur
       (dec m)
       (let [[x & xst] q]
         (-> xst
             (unique-insert (* 2 x))
             (unique-insert (* 3 x))
             (unique-insert (* 5 x))))))))
;;three queues
(defn rns3
  {:test (fn []
           (assert
            (=
             859963392
             (time (rns3 1500)))))}
  [n]
  (if (= 1 n)
    1
    (loop [m (dec n)
           q1 [2]
           q2 [3]
           q3 [5]]
      (let [x (apply min (map first [q1 q2 q3]))]
        (if (= m 1)
          x
          (let [[x1 & xst1] q1
                [x2 & xst2] q2
                [x3 & xst3] q3]
            (condp = x
              x1 (recur (dec m)
                        (conj (vec xst1) (* 2 x))
                        (conj q2 (* 3 x))
                        (conj q3 (* 5 x)))
              x2 (recur (dec m)
                        q1
                        (conj (vec xst2) (* 3 x))
                        (conj q3 (* 5 x)))
              x3 (recur (dec m)
                        q1
                        q2
                        (conj (vec xst3) (* 5 x))))))))))

(time (rns2 1500))
(time (rns3 1500))
